common/djangoapps/track/eventshim.py

```diff
@@ -0,0 +1,380 @@
+"""
+EventShims that create
+"""
+
+# pylint: disable=missing-docstring
+
+import json
+import logging
+
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import UsageKey
+
+log = logging.getLogger(__name__)
+
+
+class DottedPathMapping(object):
+    """
+    Dictionary-like object for creating keys of dotted paths.
+
+    If a key is created that ends with a dot, it will be treated as a path
+    prefix.  Any value whose prefix matches the dotted path can be used
+    as a key for that value, but only the most specific match will
+    be used.
+    """
+
+    def __init__(self, registry=None):
+        self._match_registry = {}
+        self._prefix_registry = {}
```

Comment:
Partly for simplicity, partly for performance reasons.  

1. With a regex-based solution, every lookup would require: a linear number of lookups, each of which could be slow.
2. It would be prohibitively hard to determine which is the most specific match.  
3. The current implementation is linear for prefix matches, but constant for exact matches, but the prefixed lookup can be refactored to a O(log N) lookup in a tree without changing the interface at all.  (If you want me to do this I can.  If you want me to leave a note that this should be done if the number of shims increases, I can do that too).
5. Regexes provide an interface with a large surface area, most of which would complicate rather than simplify this problem (named matches, lookahead, backtracking, etc.)

They just felt like the wrong tool for this particular job.